### ISSUE TITLE
CDI EJB Field producers' validation fail with incompatible types

### ISSUE THREAD

**Issue by pdudits** [payara/Payara#668]
I'm deploying an EAR, within which there is a war with producer:

````
public class ServiceProducer {
    @EJB(lookup = "java:/app/ejb-module/EjbInterfaceImpl")
    @Produces
    @AlternativeStereotype // stereotype enabled in beans.xml
    EjbInterface contractService;
}
````

Deployment fails with
````
[2016-02-19 17:14:17,673] [admin-listener(1)] [ERROR] [javax.enterprise.system.core]: Exception while loading the app : CDI definition failure:The type of the injection point fieldName is EjbInterface.  The type of the physical resource is com.pletely.DifferentEjb They are incompatible. 
javax.enterprise.inject.spi.DefinitionException: The type of the injection point fieldName is EjbInterface.  The type of the physical resource is com.pletely.DifferentEjb They are incompatible. 
	at org.glassfish.weld.services.InjectionServicesImpl.throwProducerDefinitionExeption(InjectionServicesImpl.java:319) ~[weld-integration.jar:na]
	at org.glassfish.weld.services.InjectionServicesImpl.validateResourceClass(InjectionServicesImpl.java:310) ~[weld-integration.jar:na]
	at org.glassfish.weld.services.InjectionServicesImpl.validateEjbProducer(InjectionServicesImpl.java:234) ~[weld-integration.jar:na]
	at org.glassfish.weld.services.InjectionServicesImpl.registerInjectionTarget(InjectionServicesImpl.java:197) ~[weld-integration.jar:na]
	at org.jboss.weld.manager.InjectionTargetFactoryImpl.postProcess(InjectionTargetFactoryImpl.java:156) ~[weld-osgi-bundle.jar:2015-09-08 14:57]
	at org.jboss.weld.manager.InjectionTargetFactoryImpl.createInjectionTarget(InjectionTargetFactoryImpl.java:93) ~[weld-osgi-bundle.jar:2015-09-08 14:57]
	at org.jboss.weld.bean.ManagedBean.<init>(ManagedBean.java:100) ~[weld-osgi-bundle.jar:2015-09-08 14:57]
	at org.jboss.weld.bean.ManagedBean.of(ManagedBean.java:80) ~[weld-osgi-bundle.jar:2015-09-08 14:57]
	at org.jboss.weld.bootstrap.AbstractBeanDeployer.createManagedBean(AbstractBeanDeployer.java:264) ~[weld-osgi-bundle.jar:2015-09-08 14:57]
	at org.jboss.weld.bootstrap.BeanDeployer.createClassBean(BeanDeployer.java:228) ~[weld-osgi-bundle.jar:2015-09-08 14:57]
	at org.jboss.weld.bootstrap.BeanDeployer.createClassBeans(BeanDeployer.java:191) ~[weld-osgi-bundle.jar:2015-09-08 14:57]
	at org.jboss.weld.bootstrap.BeanDeployment.createBeans(BeanDeployment.java:256) ~[weld-osgi-bundle.jar:2015-09-08 14:57]
	at org.jboss.weld.bootstrap.WeldStartup.deployBeans(WeldStartup.java:394) ~[weld-osgi-bundle.jar:2015-09-08 14:57]
	at org.jboss.weld.bootstrap.WeldBootstrap.deployBeans(WeldBootstrap.java:83) ~[weld-osgi-bundle.jar:2015-09-08 14:57]
	at org.glassfish.weld.WeldDeployer.event(WeldDeployer.java:222) ~[weld-integration.jar:na]
	at org.glassfish.kernel.event.EventsImpl.send(EventsImpl.java:131) ~[kernel.jar:na]
	at org.glassfish.internal.data.ApplicationInfo.load(ApplicationInfo.java:329) ~[internal-api.jar:na]
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:496) ~[kernel.jar:na]
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:219) ~[kernel.jar:na]
````

When debugging, at `validateEjbProducer(InjectionServicesImpl.java:234)` all elements of ejbs Collection have null jndiName. Therefore `oneEjb.getJndiName(`) returns empty string and that matches any `lookupName`, so first ejb in the list is elected as a match.

**Comment by pdudits** [comment]
The workaround is to use producer method, which are not validated at runtime.

It also could be related to #391 - I've seen more places where things fail in case of previously failed deployment. But restarting the server was so far much quicker solution than finding why Descriptor Bundles from different managers differ.

**Comment by pdudits** [comment]
I've hit this issue again, this time my test was deploying a simple .war that imports an external EJB. So neither EAR packaging or stereotype are relevant for the issue. I'll try to spend a little time creating a sample WAR file.

**Comment by mikecroft** [comment]
thanks @pdudits - do you still think this is related to #391? I've had a quick look at reproducing and couldn't, so a sample would be very useful

**Comment by pdudits** [comment]
On second occasion I haven't had a previously failed deployment so #391 is likely not related. 

I managed to create a test case, I'll push it to GitHub after the weekend.

**Comment by pdudits** [comment]
The test case is available here: https://github.com/pdudits/payara-testcases/tree/master/issue-668

First deploy `ejb-module`, deployment of `web-module` will fail, at least on JDK8.

**Comment by mikecroft** [comment]
Excellent, thanks very much!

**Comment by lprimak** [comment]
Could this be related to #315 ?  It's been recently fixed.

**Comment by pdudits** [comment]
The testcase fails immediately at first deployment, so it's not related to dirty application bundle.

**Comment by mikecroft** [comment]
Verified and created internal issue `PAYARA-947`

**Comment by lprimak** [comment]
Fixed by PR #1017
